packages/belos/src/BelosGmresPolyOp.hpp

```diff
@@ -100,31 +267,425 @@ namespace Belos {
+    gmres_iter->initializeGmres(newstate);
+
+    // Perform Gmres iteration
+    try {
+      gmres_iter->iterate();
+    }
+    catch (GmresIterationOrthoFailure e) {
+      // Try to recover the most recent least-squares solution
+      gmres_iter->updateLSQR( gmres_iter->getCurSubspaceDim() );
+    }
+    catch (std::exception e) {
+      using std::endl;
+      printer_->stream(Errors) << "Error! Caught exception in BlockGmresIter::iterate() at iteration "
+        << gmres_iter->getNumIters() << endl << e.what () << endl;
+      throw;
+    }
+
+    // Get the solution for this polynomial, use in comparison below
+    Teuchos::RCP<MV> currX = gmres_iter->getCurrentUpdate();
+  
+    // Record polynomial info, get current GMRES state
+    GmresIterationState<ScalarType,MV> gmresState = gmres_iter->getState();
+  
+    // If the polynomial has no dimension, the tolerance is too low, return false
+    dim_ = gmresState.curDim;
+    if (dim_ == 0) {
+      return;
+    }
+ 
+    //  Make a view and then copy the RHS of the least squares problem.
+    //
+    y_ = Teuchos::SerialDenseMatrix<OT,ScalarType>( Teuchos::Copy, *gmresState.z, dim_, 1 );
+    H_ = *gmresState.H;
+
+    //
+    // Solve the least squares problem.
+    //
+    Teuchos::BLAS<OT,ScalarType> blas;
+    blas.TRSM( Teuchos::LEFT_SIDE, Teuchos::UPPER_TRI, Teuchos::NO_TRANS,
+               Teuchos::NON_UNIT_DIAG, dim_, 1, SCT::one(),
+               gmresState.R->values(), gmresState.R->stride(),
+               y_.values(), y_.stride() );
+
+  }
+
+  template <class ScalarType, class MV, class OP>
+  void GmresPolyOp<ScalarType, MV, OP>::ApplyPoly( const MV& x, MV& y ) const 
+  {
+    if (dim_) {
+      if (polyType_ == "Arnoldi")
+        ApplyArnoldiPoly(x, y);
+      else if (polyType_ == "Gmres")
+        ApplyGmresPoly(x, y);
+    }
+    else {
+      // Just apply the operator in problem_ to x and return y.
+      problem_->applyOp( x, y );
+    }  
+  }
+
+  template <class ScalarType, class MV, class OP>
+  void GmresPolyOp<ScalarType, MV, OP>::ApplyGmresPoly( const MV& x, MV& y ) const 
+  {
+    Teuchos::RCP<MV> AX = MVT::CloneCopy(x);
+    Teuchos::RCP<MV> AX2 = MVT::Clone( x, MVT::GetNumberVecs(x) );
+
+    // Apply left preconditioner.
+    if (!LP_.is_null()) {
+      Teuchos::RCP<MV> Xtmp = MVT::Clone( x, MVT::GetNumberVecs(x) );
+      problem_->applyLeftPrec( *AX, *Xtmp ); // Left precondition x into the first vector 
+      AX = Xtmp;
+    } 
+
+    MVT::MvAddMv(pCoeff_(0,0), *AX, SCT::zero(), y, y); //y= coeff_i(A^ix)
+    for( int i=1; i < dim_+1; i++)
+    {
+      Teuchos::RCP<MV> X, Y;
+      if ( i%2 )
+      {
+        X = AX;
+        Y = AX2;
+      }
+      else
+      {
+        X = AX2;
+        Y = AX;
+      }
+      problem_->apply(*X, *Y);
+      MVT::MvAddMv(pCoeff_(i,0), *Y, SCT::one(), y, y); //y= coeff_i(A^ix) +y
+    }
+
+    // Apply right preconditioner.
+    if (!RP_.is_null()) {
+      Teuchos::RCP<MV> Ytmp = MVT::CloneCopy(y);
```

Comment:
Might want to think about having Ytmp as an instance variable, or at least hoisting its creation out of this loop into solver initialization.